proposals/0000-class-method-headers.rst

```diff
@@ -0,0 +1,437 @@
+(visible forall in types of terms). If we want to visibly quantify over some of
+class variables, there are also two workarounds:
+
+* Dummy type variable with an equality constraint:
+  ::
+
+    class C a b c where
+      f :: forall c' -> (c ~ c') => (a, b, c) -> Bool
+
+* Again, a wrapper function:
+  ::
+
+    fWrapper :: forall b. forall c -> forall a. C a b c => (a, b, c) -> Bool
+    fWrapper @b c @a = f @a @b @c
+
+Our proposal is to take the LHS of the wrapper, ``fWrapper @b c @a``, and
+integrate it into the original ``f``'s declaration.
+
+This approach avoids the main problem of the wrapper workaround: the
+introduction of unnecessary named entities to the APIs of libraries (it is
+generally desirable to keep APIs simple and easy to browse).
+
+Proposed Change Specification
+-----------------------------
+
+Syntax
+~~~~~~
+
+Take the Haskell 2010 class method signature grammar as the starting point::
+
+  gendecl -> vars :: [context =>] type
+  vars    -> var_1 , ... , var_n         (n ≥ 1)
+
+Instead of variable names for class method left-hand sides, we introduce the
+notion of a signature header::
+
+  sighdr  -> var (sigbndr_1 ... sigbndr_n)
+  sigbndr -> tyvar
+           | @tyvar
+
+  gendecl -> sighdrs :: [context =>] type
+  sighdrs -> sighdr_1 , ... , sighdr_n         (n ≥ 1)
+
+A validity check ensures that the binders are only used in class method
+signatures and are disallowed in function signatures.
+
+The syntax of function bindings, including method definitions in instances, is
+assumed to be extended by `"Type variable binders"
+<https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0155-type-lambda.rst>`_
+to allow type binders prefixed with ``@``::
+
+  f @Int x = ...
+
+Semantics
+~~~~~~~~~
+
+1. When one or more binders are present in a class method signature, we consider
+   it a class method header, and require the ``-XClassMethodHeaders`` extension.
+
+#. A class method header, if present, must bind every class variable mentioned
+   in the class declaration header, and must bind it exactly once.
+
+#. No other variables can be bound in the class method header.
+
+#. In the compiler-generated top-level signature for the class method, variables
+   bound as ``@a`` are quantified with ``forall a.``, and variables bound as
+   ``a`` are quantified with ``forall a ->``.
+
+#. In the compiler-generated top-level signature for the class method, the order
+   of quantifiers matches the order of binders in the class method header.
+
+#. The variables may not be quantified out of dependency order.
+
+#. In instances, method definitions *may* explicitly write out instantiations of
+   invisible class variables::
+
+     class C a where
+       f @a :: a
+
+     instance C Int where
+       f @Int = 42
+
+   The type in the method equation must be identical to the one in the instance
+   head.
+
+#. In instances, method definitions *must* explicitly write out
+   instantiations of visible class variables::
```

Comment:
Fair enough. But, unless I'm mistaken, it doesn't look like requiring visible arguments to be specified is strictly necessary. Would you discuss the reason for this choice in the _Alternatives_ section? It would be helpful